src/org/scratchOrgCreate.ts

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+// third
+import { Duration } from '@salesforce/kit';
+import { ensureString, getString } from '@salesforce/ts-types';
+
+// Local
+import { Messages } from '../messages';
+import { Logger } from '../logger';
+import { ConfigAggregator } from '../config/configAggregator';
+import { SfdxProject } from '../sfdxProject';
+import { SfdxError } from '../sfdxError';
+import { Org } from './org';
+import {
+  authorizeScratchOrg,
+  requestScratchOrgCreation,
+  pollForScratchOrgInfo,
+  deploySettingsAndResolveUrl,
+  ScratchOrgInfo,
+} from './scratchOrgInfoApi';
+import SettingsGenerator from './scratchOrgSettingsGenerator';
+import { generateScratchOrgInfo, getScratchOrgInfoPayload } from './scratchOrgInfoGenerator';
+import { AuthFields, AuthInfo } from './authInfo';
+import { Connection } from './connection';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.load('@salesforce/core', 'scratchOrgCreate', ['SourceStatusResetFailure']);
+
+export const DEFAULT_STREAM_TIMEOUT_MINUTES = 6;
+
+export interface ScratchOrgCreateResult {
+  username?: string;
+  scratchOrgInfo?: ScratchOrgInfo;
+  authInfo?: AuthInfo;
+  authFields?: AuthFields;
+  warnings: string[];
+}
+
+/**
+ * interface ScratchOrgCreateOptions
+ *
+ * @param hubOrg the environment hub org
+ * @param connectedAppConsumerKey The connected app consumer key.
+ * @param durationDays duration of the scratch org (in days) (default:1, min:1, max:30)
+ * @param nonamespace create the scratch org with no namespace
+ * @param noancestors do not include second-generation package ancestors in the scratch org
+ * @param wait the streaming client socket timeout (in minutes) must be an instance of the Duration utility class (default:6, min:2)
+ * @param retry number of scratch org auth retries after scratch org is successfully signed up (default:0, min:0, max:10)
+ * @param apiversion target server instance API version
+ * @param definitionjson org definition in JSON format
+ * @param definitionfile path to an org definition file
+ * @param orgConfig overrides definitionjson
+ * @param clientSecret OAuth client secret of personal connected app
+ */
+
+export interface ScratchOrgCreateOptions {
+  hubOrg: Org;
+  connectedAppConsumerKey?: string;
+  durationDays?: number;
+  nonamespace?: boolean;
+  noancestors?: boolean;
+  wait?: Duration;
+  retry?: number;
+  apiversion?: string;
+  definitionjson?: string;
+  definitionfile?: string;
+  orgConfig?: Record<string, unknown>;
+  clientSecret?: string;
+}
+
+const validateDuration = (durationDays: number): void => {
+  const min = 1;
+  const max = 30;
+  if (Number.isInteger(durationDays)) {
+    if (durationDays < min) {
+      throw new SfdxError(
```

Comment:
It would be nice to have the exception messages moved to message file and then use messages.createError(...) to create the SfdxError instance to throw.